### ISSUE TITLE
com.hazelcast.internal.serialization.impl.compact.schema.CompactSchemaReplicationSlowTest [API-1994]

### ISSUE THREAD

**Issue by olukas** [hazelcast/hazelcast#23468]
_master_ (commit 74134e54b476ce07f4cf47c8682b43d9090e573f)

Failed on zing-8: https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-ZingJDK8-nightly/15/testReport/com.hazelcast.internal.serialization.impl.compact.schema/CompactSchemaReplicationSlowTest/

Following tests failed:
- `com.hazelcast.internal.serialization.impl.compact.schema.CompactSchemaReplicationSlowTest.testSchemaReplication_whenAMemberThrowsRetryableExceptionAllTheTime_duringAcknowledgmentPhase`
- `com.hazelcast.internal.serialization.impl.compact.schema.CompactSchemaReplicationSlowTest.testSchemaReplication_whenAMemberThrowsRetryableExceptionAllTheTime_duringPreparationPhase`

<details><summary>Stacktrace:</summary>

```
java.lang.NoClassDefFoundError: Could not initialize class com.hazelcast.internal.serialization.impl.compact.schema.CompactSchemaReplicationSlowTest
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.junit.runners.BlockJUnit4ClassRunner.createTest(BlockJUnit4ClassRunner.java:250)
	at com.hazelcast.test.AbstractParameterizedHazelcastClassRunner.createTest(AbstractParameterizedHazelcastClassRunner.java:87)
	at org.junit.runners.BlockJUnit4ClassRunner.createTest(BlockJUnit4ClassRunner.java:260)
	at org.junit.runners.BlockJUnit4ClassRunner$2.runReflectiveCall(BlockJUnit4ClassRunner.java:309)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.BlockJUnit4ClassRunner.methodBlock(BlockJUnit4ClassRunner.java:306)
	at com.hazelcast.test.AbstractHazelcastClassRunner.methodBlock(AbstractHazelcastClassRunner.java:228)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at com.hazelcast.test.HazelcastParallelClassRunner.access$301(HazelcastParallelClassRunner.java:49)
	at com.hazelcast.test.HazelcastParallelClassRunner$TestRunner.run(HazelcastParallelClassRunner.java:162)
	at java.lang.Thread.run(Thread.java:807)

```
</details>

<details><summary>Standard output:</summary>

```
Finished Running Test: testSchemaReplication_whenAMemberThrowsRetryableExceptionAllTheTime_duringPreparationPhase in 0.428 seconds.
Finished Running Test: testSchemaReplication_whenAMemberThrowsRetryableExceptionAllTheTime_duringAcknowledgmentPhase in 0.432 seconds.

```
</details>

Standard output can be found here - https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-master-ZingJDK8-nightly/15/

**Comment by olukas** [comment]
Failed again on _master_ with zing-8: https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-ZingJDK8-nightly/18/testReport/com.hazelcast.internal.serialization.impl.compact.schema/CompactSchemaReplicationSlowTest/

**Comment by sumnerib** [comment]
Failed again on _master_ with zing-8: https://jenkins.hazelcast.com/job/Hazelcast-master-ZingJDK8-nightly/22/testReport/junit/com.hazelcast.internal.serialization.impl.compact.schema/CompactSchemaReplicationSlowTest/

**Comment by olukas** [comment]
Failed again on _master_ with zing-8: https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-ZingJDK8-nightly/23/testReport/com.hazelcast.internal.serialization.impl.compact.schema/CompactSchemaReplicationSlowTest/

**Comment by Patras3** [comment]
Failed again on _master_ with zing-8: https://jenkins.hazelcast.com/job/Hazelcast-master-ZingJDK8-nightly/24/testReport/junit/com.hazelcast.internal.serialization.impl.compact.schema/CompactSchemaReplicationSlowTest

**Comment by olukas** [comment]
Failed again on _master_ with zing-8: https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-ZingJDK8-nightly/25/testReport/com.hazelcast.internal.serialization.impl.compact.schema/CompactSchemaReplicationSlowTest/

**Comment by mdumandag** [comment]
This is the same bug as the https://github.com/hazelcast/hazelcast/issues/23477. It started failing after mockito upgrade, and fails frequently on zing JDK 8. 

The failures are the same, mockito is trying to call a real method of the mocked object and fails with NPE. Here is the log from the first failure:

```
[ERROR] com.hazelcast.internal.serialization.impl.compact.schema.CompactSchemaReplicationSlowTest.testSchemaReplication_whenAMemberThrowsRetryableExceptionAllTheTime_duringAcknowledgmentPhase  Time elapsed: 0.419 s  <<< ERROR!
java.lang.ExceptionInInitializerError
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.junit.runners.BlockJUnit4ClassRunner.createTest(BlockJUnit4ClassRunner.java:250)
	at com.hazelcast.test.AbstractParameterizedHazelcastClassRunner.createTest(AbstractParameterizedHazelcastClassRunner.java:87)
	at org.junit.runners.BlockJUnit4ClassRunner.createTest(BlockJUnit4ClassRunner.java:260)
	at org.junit.runners.BlockJUnit4ClassRunner$2.runReflectiveCall(BlockJUnit4ClassRunner.java:309)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.BlockJUnit4ClassRunner.methodBlock(BlockJUnit4ClassRunner.java:306)
	at com.hazelcast.test.AbstractHazelcastClassRunner.methodBlock(AbstractHazelcastClassRunner.java:228)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at com.hazelcast.test.HazelcastParallelClassRunner.access$301(HazelcastParallelClassRunner.java:49)
	at com.hazelcast.test.HazelcastParallelClassRunner$TestRunner.run(HazelcastParallelClassRunner.java:162)
	at java.lang.Thread.run(Thread.java:807)
Caused by: java.lang.RuntimeException: java.lang.NullPointerException
	at com.hazelcast.internal.serialization.impl.compact.CompactTestUtil.getSchemasFor(CompactTestUtil.java:349)
	at com.hazelcast.internal.serialization.impl.compact.schema.CompactSchemaReplicationTestBase.<clinit>(CompactSchemaReplicationTestBase.java:51)
	... 17 more
Caused by: java.lang.NullPointerException
	at com.hazelcast.internal.serialization.impl.compact.CompactStreamSerializer.canBeSerializedAsCompact(CompactStreamSerializer.java:95)
	at com.hazelcast.internal.serialization.impl.compact.CompactUtil.verifyFieldClassShouldBeSerializedAsCompact(CompactUtil.java:203)
	at com.hazelcast.internal.serialization.impl.compact.zeroconfig.ValueReaderWriters.readerWriterFor(ValueReaderWriters.java:215)
	at com.hazelcast.internal.serialization.impl.compact.ReflectiveCompactSerializer.createFastReadWriteCaches(ReflectiveCompactSerializer.java:299)
	at com.hazelcast.internal.serialization.impl.compact.ReflectiveCompactSerializer.write(ReflectiveCompactSerializer.java:82)
	at com.hazelcast.internal.serialization.impl.compact.CompactTestUtil.getSchemasFor(CompactTestUtil.java:347)
	... 18 more
```

where the NPE throwing method is in fact stubbed (CompactTestUtil#getSchemasFor)

```
...
        CompactStreamSerializer compactStreamSerializer = mock(CompactStreamSerializer.class);
        when(compactStreamSerializer.canBeSerializedAsCompact(any())).thenReturn(true);
...
```

So, I will not do anything on this issue for now.

**Comment by mdumandag** [comment]
PRs to downgrade mockito have been merged into OS and Enterprise repos recently. I will close this issue for now, and we can re-open and re-investigate if it fails again despite the downgrade.

**Comment by olukas** [comment]
Reopening as issue occurred again on: _master_ (commit ba36217314220bbd9fa955ab82b7e4a4c9adb51b)
with zing-8: https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-ZingJDK8-nightly/27/testReport/com.hazelcast.internal.serialization.impl.compact.schema/CompactSchemaReplicationSlowTest/testSchemaReplication_whenAMemberThrowsRetryableExceptionAllTheTime_duringPreparationPhase/
<details><summary>Stacktrace:</summary>

```
java.lang.NoClassDefFoundError: Could not initialize class com.hazelcast.internal.serialization.impl.compact.schema.CompactSchemaReplicationSlowTest
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.junit.runners.BlockJUnit4ClassRunner.createTest(BlockJUnit4ClassRunner.java:250)
	at com.hazelcast.test.AbstractParameterizedHazelcastClassRunner.createTest(AbstractParameterizedHazelcastClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.createTest(BlockJUnit4ClassRunner.java:260)
	at org.junit.runners.BlockJUnit4ClassRunner$2.runReflectiveCall(BlockJUnit4ClassRunner.java:309)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.BlockJUnit4ClassRunner.methodBlock(BlockJUnit4ClassRunner.java:306)
	at com.hazelcast.test.AbstractHazelcastClassRunner.methodBlock(AbstractHazelcastClassRunner.java:228)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at com.hazelcast.test.HazelcastParallelClassRunner.access$301(HazelcastParallelClassRunner.java:49)
	at com.hazelcast.test.HazelcastParallelClassRunner$TestRunner.run(HazelcastParallelClassRunner.java:162)
	at java.lang.Thread.run(Thread.java:807)

```
</details>


<details><summary>Standard output:</summary>

```
Finished Running Test: testSchemaReplication_whenAMemberThrowsRetryableExceptionAllTheTime_duringPreparationPhase in 0.448 seconds.
Finished Running Test: testSchemaReplication_whenAMemberThrowsRetryableExceptionAllTheTime_duringAcknowledgmentPhase in 0.449 seconds.

```
</details>

Standard output can be found here - https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-master-ZingJDK8-nightly/27/

**Comment by olukas** [comment]
Failed again on _master_ with zing-8: https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-ZingJDK8-nightly/29/testReport/com.hazelcast.internal.serialization.impl.compact.schema/CompactSchemaReplicationSlowTest/

**Comment by JamesHazelcast** [comment]
Unfortunately the S3 logs don't show much useful information for this failure, and Jenkins logs have now expired, but as comments above show this failure is related to Mockito - we've had a number of Mockito-related issues in recent weeks and months, with several versions trialed. On May 24th we upgraded to Mockito v5.3.1, specifically to try and address failures such as this - it's worth noting that this test has not failed since then.

In some test cases that were failing due to Mockito, we replaced the mocked portions of the test with custom implementations - if this test failure occurs again then I suggest we do the same for this test. However, since such a change would be slightly more complicated than other such cases, I think it's best we wait and see if this failure is reproduced again first. All failures occurred on JDK 8 only (which we have dropped mainstream support for with 5.4), and hopefully our Mockito version bump is enough to stave on further failures on maintenance jobs.

Closing this issue for now - if it surfaces again due to Mockito, I suggest we replace the mocked implementation with a simple alternative that avoids Mockito.
